File: services/api/src/nova/schemas/analytics.py
"""Analytics payloads.

Every response carries its ``coverage`` block. A chart without one invites
the reader to treat two days of data the same as two months, and that is the
mistake this whole phase is built to avoid.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

# Every request is a scan over telemetry, so the window is capped. A bound
# rather than a closed set of allowed values: the bound is the property that
# actually matters, and an enum of four magic numbers would only mean clients
# could not ask for the fortnight they wanted.
MIN_WINDOW_DAYS = 1
MAX_WINDOW_DAYS = 90
DEFAULT_WINDOW_DAYS = 30


class CoverageRead(BaseModel):
    """How much evidence the numbers below rest on."""

    total_events: int
    distinct_days: int
    first_event_at: datetime | None
    last_event_at: datetime | None
    # Whether there is enough here to say anything about habits. The app
    # shows a different screen when this is false, rather than charts of
    # three data points.
    is_sufficient: bool


class HourBucket(BaseModel):
    """One local hour of the day."""

    hour: int = Field(ge=0, le=23)
    count: int


class DayBucket(BaseModel):
    day: date
    count: int


class WeekdayHourBucket(BaseModel):
    """One cell of the week grid. 0 is Monday, matching Python."""

    weekday: int = Field(ge=0, le=6)
    hour: int = Field(ge=0, le=23)
    count: int


class BatteryPoint(BaseModel):
    recorded_at: datetime
    percent: int = Field(ge=0, le=100)


class EventTypeBucket(BaseModel):
    event_type: str
    count: int


class GapRead(BaseModel):
    """A silence between heartbeats."""

    started_at: datetime
    ended_at: datetime
    seconds: int


class AnalyticsRead(BaseModel):
    """Everything the insights screen plots."""

    window_days: int
    # The timezone every bucket below was computed in, echoed so a client
    # rendering "14:00" knows whose two o'clock it is.
    timezone: str
    coverage: CoverageRead

    presence_by_hour: list[HourBucket]
    presence_by_day: list[DayBucket]
    presence_by_weekday_hour: list[WeekdayHourBucket]
    battery: list[BatteryPoint]
    event_types: list[EventTypeBucket]
    gaps: list[GapRead]


class InsightRead(BaseModel):
    """One statement, with the evidence behind it."""

    kind: str
    headline: str
    detail: str
    confidence: Literal["low", "medium", "high"]
    sample_size: int
    days_observed: int


class InsightsRead(BaseModel):
    """What can be said, and what cannot be said yet."""

    window_days: int
    timezone: str
    coverage: CoverageRead
    insights: list[InsightRead]
    # Present when there is too little data to say anything. Explicit, so a
    # client shows "still getting to know you" rather than an empty list that
    # is indistinguishable from a bug.
    insufficient_reason: str | None = None

File: apps/ios/NOVA/Models/Analytics.swift
import Foundation

/// How much evidence the numbers rest on.
///
/// Carried on every analytics response and shown on screen. A chart without
/// it invites the reader to treat two days of data the same as two months.
struct Coverage: Codable, Equatable, Sendable {
    let totalEvents: Int
    let distinctDays: Int
    let firstEventAt: Date?
    let lastEventAt: Date?
    /// Whether there is enough here to say anything about habits.
    let isSufficient: Bool
}

/// One local hour of the day.
struct HourBucket: Codable, Identifiable, Equatable, Sendable {
    let hour: Int
    let count: Int

    var id: Int { hour }

    /// "09:00". Zero-padded so the axis labels line up.
    var label: String { String(format: "%02d:00", hour) }
}

struct DayBucket: Codable, Identifiable, Equatable, Sendable {
    let day: Date
    let count: Int

    var id: Date { day }
}

/// One cell of the 7×24 week grid. 0 is Monday, matching the server.
struct WeekdayHourBucket: Codable, Identifiable, Equatable, Sendable {
    let weekday: Int
    let hour: Int
    let count: Int

    var id: Int { weekday * 24 + hour }
}

struct BatteryPoint: Codable, Identifiable, Equatable, Sendable {
    let recordedAt: Date
    let percent: Int

    var id: Date { recordedAt }
}

struct EventTypeBucket: Codable, Identifiable, Equatable, Sendable {
    let eventType: String
    let count: Int

    var id: String { eventType }

    /// "person_detected" reads as "Person detected".
    var label: String {
        eventType.replacingOccurrences(of: "_", with: " ").capitalizedFirst
    }
}

/// A silence between heartbeats.
struct Gap: Codable, Identifiable, Equatable, Sendable {
    let startedAt: Date
    let endedAt: Date
    let seconds: Int

    var id: Date { startedAt }
}

/// Everything the insights screen plots.
struct Analytics: Codable, Equatable, Sendable {
    let windowDays: Int
    /// The timezone every bucket was computed in. Echoed by the server so a
    /// client rendering "14:00" knows whose two o'clock it is.
    let timezone: String
    let coverage: Coverage

    let presenceByHour: [HourBucket]
    let presenceByDay: [DayBucket]
    let presenceByWeekdayHour: [WeekdayHourBucket]
    let battery: [BatteryPoint]
    let eventTypes: [EventTypeBucket]
    let gaps: [Gap]

    /// The busiest cell, used to scale the heatmap. 1 rather than 0 when
    /// there is nothing, so nothing divides by zero.
    var peakWeekdayHourCount: Int {
        max(presenceByWeekdayHour.map(\.count).max() ?? 0, 1)
    }
}

/// One statement, with the evidence behind it.
struct Insight: Codable, Identifiable, Equatable, Sendable {
    enum Confidence: String, Codable, Sendable {
        case low, medium, high

        /// Shown next to the headline. A reader deciding whether to believe
        /// a claim about their own life deserves to see this.
        var label: String {
            switch self {
            case .low: "Tentative"
            case .medium: "Fairly sure"
            case .high: "Confident"
            }
        }
    }

    let kind: String
    let headline: String
    let detail: String
    let confidence: Confidence
    let sampleSize: Int
    let daysObserved: Int

    var id: String { kind }

    var symbol: String {
        switch kind {
        case "active_hours": "clock"
        case "busiest_weekday": "calendar"
        case "weekly_trend": "chart.line.uptrend.xyaxis"
        case "battery_runtime": "battery.50"
        case "connectivity": "wifi.exclamationmark"
        default: "sparkles"
        }
    }

    /// The support, spelled out. Deliberately not hidden behind a tap.
    var evidence: String {
        "\(sampleSize) events over \(daysObserved) day\(daysObserved == 1 ? "" : "s")"
    }
}

/// What can be said, and what cannot be said yet.
struct Insights: Codable, Equatable, Sendable {
    let windowDays: Int
    let timezone: String
    let coverage: Coverage
    let insights: [Insight]
    /// Set when NOVA declines to draw conclusions. An empty list *with* a
    /// reason is a normal response; an empty list without one would be
    /// indistinguishable from a bug.
    let insufficientReason: String?
}

private extension String {
    var capitalizedFirst: String {
        guard let first else { return self }
        return first.uppercased() + dropFirst()
    }
}
